docs/source/scheduler.rst

```diff
@@ -0,0 +1,263 @@
+orders all application (per-partition) into a list. The list is ordered by
+"utilization", which is a heuristic that takes into account the capacity
+reserved by a given tenant, accumulated capacity demand, and other factors. The
+exact details of application ordering will be described in detail later in the
+document.
+
+Once the application queue is constructed, Treadmill scheduler will try and
+find suitable placement for each application, such that application demand and
+other constraints are satisfied. The placement is "online", once the placement
+is found, it is not revisited within the given scheduling cycle. Applications
+on the left of the queue has strict priority over application on the right of
+the queue. If a placement cannot be found, the scheduler will try to find
+capacity by evicting applications from the right end of the queue. For those
+applications that were evicted the scheduler will try to restore the placement
+or find new one. If this is not possible, application will be evicted, yet it
+will be left on the placement queue in pending state.
+
+Once the queue evaluation is done, the scheduler will publish final placement
+and wait for new events.
+
+
+.. _section-utilization:
+
+Tenants, allocations, and application utilization
+-------------------------------------------------
+
+Treadmill scheduler supports multi tenancy by assigning to each tenant one or
+more allocations. Administratively, allocations are protected resources, and
+Treadmill API will enforce access control on allocations. From scheduler
+perspective the mechanics of allocation entitlements are not important.
+
+Allocation is used to reserve capacity on a given partition.
+
+ * Allocation has associated capacity (e.g memory, cpu, disk) and rank.
+ * Each application is assigned to a single allocation. Applications within a
+   given allocation form a "private allocation queue".
+ * Allocation has additional property called "rank" and "rank adjustment",
+   which are important when constructing global application demand queue.
+ * Allocations form a hierarchy. Each allocation can have child allocations,
+   with their own associated capacity, application queue and rank.
+
+Allocation private queue is controlled by the user. Within the private
+allocation queue, applications are ordered by "priority". Users have full
+control over specifying application priority, and application priority can be
+adjusted anytime by the application owner.
+
+Priority allows individual users to express which applications are more
+important for them. Priority values range from 100 (most important) to 0.
+Priorities matter only within a given allocation.
+
+Advanced: Priority 0 has special meaning. Setting application priority to zero
+for instruct the scheduler to place this application at the very end of the
+global queue. As such, priority zero applications will run only when there is
+no capacity pressure on the cell.
```

Comment:
--- They're probably also highly likely to be evicted -e .g. they're like "spot" compute jobs, of sorts, right?

Reply:
Moved content to priority-0 section below. Clarified that section